### DISCUSSION TITLE
Add custom query

### DISCUSSION THREAD

**Discussion by bad1987** [pimcore/pimcore#13530] [Q&A / Support] [answered]
Hi everyone!
I'm trying to a custom query by following this documentation: https://pimcore.com/docs/data-hub/current/GraphQL/Query/Add_Custom_Query.html. My problem is that I tried to reproduce the example given in the doc but I don't know in which file should I define my custom query. I tried with event subscriber but it didn't work. Any suggestion would be helpful. This is my code:

`<?php
class GraphQlSubscriber implements EventSubscriberInterface
{
    public static function getSubscribedEvents()
    {
        return [
            \Pimcore\Bundle\DataHubBundle\Event\GraphQL\QueryEvents::PRE_BUILD => 'onQueryPreEvent'
        ];
    }
    public function onQueryPreEvent(\Pimcore\Bundle\DataHubBundle\Event\GraphQL\Model\QueryTypeEvent $event) {
        $config = $event->getConfig();

        $outputType = new \GraphQL\Type\Definition\ObjectType([
                'name' => "CustomQueryResultType",
                'fields' => [
                    'outputFieldA' => [
                        'type' => \GraphQL\Type\Definition\Type::string(),
                        'resolve' => function ($source, $args, $context, \GraphQL\Type\Definition\ResolveInfo $info) {
                            return "A-value for item " . $source['resolvedId'] . " is " . uniqid();
                        }
                    ],
                    'outputFieldB' => [
                        'type' => \GraphQL\Type\Definition\Type::string(),
                        'resolve' => function ($source, $args, $context, \GraphQL\Type\Definition\ResolveInfo $info) {
                            return "B-value for item " . $source['resolvedId'] . " is " . uniqid();
                        }
                    ]
                ]
            ]
        );

        $operation = [
            'type' => $outputType,
            'args' => ['itemId' => ['type' => \GraphQL\Type\Definition\Type::nonNull(\GraphQL\Type\Definition\Type::int())]],
            'resolve' => function ($source, $args, $context, \GraphQL\Type\Definition\ResolveInfo $info) {
                // resolve the item using the input parameters. Result will be passed
                // to the field-level resolvers
                return ['resolvedId' => $args['itemId']];
            }
        ];

        $config['fields']['performCustomQuery'] = $operation;
        $event->setConfig($config);
    }
}
`

**Comment by mattamon** [comment, marked as answer]
Hey @bad1987,

I used an event listener: https://pimcore.com/docs/pimcore/current/Development_Documentation/Extending_Pimcore/Event_API_and_Event_Manager.html

I created a new folder in my src folder called "EventListener" and added the Testlistener from the docs listed above.

Then I just replaced some code in the Testlistener like the following:

```php
<?php
namespace App\EventListener;

use Pimcore\Bundle\DataHubBundle\Event\GraphQL\Model\QueryTypeEvent;

class TestListener {

    public function onPreBuild (QueryTypeEvent $event) {
        $config = $event->getConfig();

        $outputType = new \GraphQL\Type\Definition\ObjectType([
                'name' => "CustomQueryResultType",
                'fields' => [
                    'outputFieldA' => [
                        'type' => \GraphQL\Type\Definition\Type::string(),
                        'resolve' => function ($source, $args, $context, \GraphQL\Type\Definition\ResolveInfo $info) {
                            return "A-value for item " . $source['resolvedId'] . " is " . uniqid();
                        }
                    ],
                    'outputFieldB' => [
                        'type' => \GraphQL\Type\Definition\Type::string(),
                        'resolve' => function ($source, $args, $context, \GraphQL\Type\Definition\ResolveInfo $info) {
                            return "B-value for item " . $source['resolvedId'] . " is " . uniqid();
                        }
                    ]
                ]
            ]
        );

        $operation = [
            'type' => $outputType,
            'args' => ['itemId' => ['type' => \GraphQL\Type\Definition\Type::nonNull(\GraphQL\Type\Definition\Type::int())]],
            'resolve' => function ($source, $args, $context, \GraphQL\Type\Definition\ResolveInfo $info) {
                // resolve the item using the input parameters. Result will be passed
                // to the field-level resolvers
                return ['resolvedId' => $args['itemId']];
            }
        ];

        $config['fields']['performCustomQuery'] = $operation;
        $event->setConfig($config);
    }
}
```
Finally adding the TestListener under config/services.yaml:

```yaml
    App\EventListener\TestListener:
        tags:
             - { name: kernel.event_listener, event: pimcore.datahub.graphql.query.preBuild, method: onPreBuild }
```

**Reply by bad1987** [reply]
Hey @mattamon 

Thanks for your solution I'll git it a try and see if it works for me

**Reply by mattamon** [reply]
Hopefully it resolves your problem. If it does please mark as resolved! Thanks!

**Reply by bad1987** [reply]
Hey @mattamon 

You solution is working. But there is only one thing I noted: if you have more than one configuration, your custom query will appear on all of them.
